File: parser/src/main/java/com/aaroncoplan/waterfall/parser/FileParser.java
package com.aaroncoplan.waterfall.parser;

import com.aaroncoplan.waterfall.generated.WaterfallLexer;
import com.aaroncoplan.waterfall.generated.WaterfallParser;

import java.util.List;

import org.antlr.v4.runtime.CharStream;
import org.antlr.v4.runtime.CharStreams;
import org.antlr.v4.runtime.CommonTokenStream;

public class FileParser {

    public static ParseResult parseFile(final String filePath) {
        final String fileContents = FileUtils.readFile(filePath);
        return parseCodeString(filePath, fileContents);
    }

    public static ParseResult parseCodeString(final String filePath, final String codeString) {
        final CharStream charStream = CharStreams.fromString(codeString);
        final WaterfallLexer waterfallLexer = new WaterfallLexer(charStream);
        waterfallLexer.removeErrorListeners();
        final CommonTokenStream tokenStream = new CommonTokenStream(waterfallLexer);

        final WaterfallParser waterfallParser = new WaterfallParser(tokenStream);
        waterfallParser.removeErrorListeners();
        final SyntaxErrorListener errorListener = new SyntaxErrorListener(filePath);
        waterfallParser.addErrorListener(errorListener);

        final WaterfallParser.ProgramContext programAST = waterfallParser.program();
        final List<String> syntaxErrors = errorListener.getSyntaxErrors();
        return new ParseResult(filePath, syntaxErrors, programAST);
    }
}

File: parser/src/main/kotlin/com/aaroncoplan/waterfall/parser/FileParser.kt
package com.aaroncoplan.waterfall.parser

import com.aaroncoplan.waterfall.generated.WaterfallLexer
import com.aaroncoplan.waterfall.generated.WaterfallParser
import org.antlr.v4.runtime.CharStreams
import org.antlr.v4.runtime.CommonTokenStream

object FileParser {

    @JvmStatic
    fun parseFile(filePath: String): ParseResult {
        val fileContents = FileUtils.readFile(filePath) ?: ""
        return parseCodeString(filePath, fileContents)
    }

    /**
     * `filePath` may be null for parsing in-memory code (used by tests). Diagnostics
     * just substitute an empty path in that case.
     */
    @JvmStatic
    fun parseCodeString(filePath: String?, codeString: String): ParseResult {
        val charStream = CharStreams.fromString(codeString)
        val lexer = WaterfallLexer(charStream)
        lexer.removeErrorListeners()
        val tokenStream = CommonTokenStream(lexer)

        val parser = WaterfallParser(tokenStream)
        parser.removeErrorListeners()
        val errorListener = SyntaxErrorListener(filePath ?: "")
        parser.addErrorListener(errorListener)

        val programAST = parser.program()
        return ParseResult(filePath ?: "", errorListener.getSyntaxErrors(), programAST)
    }
}
